### PR TITLE
[OSD-14155] Adding _id label to limited_support_enabled metric 

### PR DESCRIPTION
**Pull request by aliceh**
In this commit, I added an _id label to the limited_support_metric to be picked up in the observatorium.
- Moved exportClusterID() in main.go to set CLUSTER_ID var at start up before creating metrics aggregator

https://issues.redhat.com/browse/OSD-14155

### CODE REVIEW

**Comment by fahlmant** [comment]
@aliceh You can use the logic already present in the proxy controller to set uuid
https://github.com/openshift/osd-metrics-exporter/blob/main/controllers/proxy/proxy_controller.go#L52

File: pkg/metrics/aggregator.go [hunk 1]
```diff
@@ -156,11 +157,17 @@ func (a *AdoptionMetricsAggregator) SetClusterAdmin(enabled bool) {
...
 
-func (a *AdoptionMetricsAggregator) SetLimitedSupport(enabled bool) {
-	if enabled {
-		a.limitedSupport.Set(1)
+func (a *AdoptionMetricsAggregator) SetLimitedSupport(uuid string, enabled float64) {
+	if enabled == 1 {
+		a.limitedSupport.With(prometheus.Labels{
+			clusterIDLabel: uuid,
+		}).Set(1)
+
```
Comment (fahlmant) [inline pkg/metrics/aggregator.go:165]:
```suggestion
        labels := prometheus.Labels{
			clusterIDLabel: uuid,
		}

	if enabled == 1 {
		a.limitedSupport.With(labels).Set(1)

```

File: pkg/metrics/aggregator.go [hunk 2]
```diff
@@ -156,11 +157,19 @@ func (a *AdoptionMetricsAggregator) SetClusterAdmin(enabled bool) {
...
+		}
+
+	if enabled == 1 {
+		a.limitedSupport.With(labels).Set(1)
+
 	} else {
-		a.limitedSupport.Set(0)
+		a.limitedSupport.With(prometheus.Labels{
+			clusterIDLabel: uuid,
+		}).Set(0)
```
Comment (fahlmant) [inline pkg/metrics/aggregator.go:171]:
```suggestion
		a.limitedSupport.With(labels).Set(0)
```

File: controllers/limited_support/limited_support_controller.go [hunk 3]
```diff
@@ -48,6 +51,12 @@ func (r *LimitedSupportConfigMapReconciler) Reconcile(ctx context.Context, req c
 	reqLogger := log.WithValues("Request.Namespace", req.Namespace, "Request.Name", req.Name)
 	reqLogger.Info("Reconciling Limited Support ConfigMap")
 
+	uuid := os.Getenv(EnvClusterID)
+	if uuid == "" {
+		return ctrl.Result{}, fmt.Errorf("cluster ID returned as empty string")
+	}
+	r.MetricsAggregator.SetClusterID(uuid)
```
Comment (fahlmant) [inline controllers/limited_support/limited_support_controller.go:58]: This is setting a metric that's already set in another controller. This needs to be removed

File: pkg/metrics/aggregator.go [hunk 4]
```diff
@@ -156,11 +157,17 @@ func (a *AdoptionMetricsAggregator) SetClusterAdmin(enabled bool) {
 	}
 }
 
-func (a *AdoptionMetricsAggregator) SetLimitedSupport(enabled bool) {
-	if enabled {
-		a.limitedSupport.Set(1)
+func (a *AdoptionMetricsAggregator) SetLimitedSupport(uuid string, enabled float64) {
```
Comment (fahlmant) [inline pkg/metrics/aggregator.go:160]:
```suggestion
func (a *AdoptionMetricsAggregator) SetLimitedSupport(uuid string, enabled bool) {
```

File: pkg/metrics/aggregator.go [hunk 5]
```diff
@@ -90,7 +91,7 @@ func NewMetricsAggregator(aggregationInterval time.Duration) *AdoptionMetricsAgg
 		aggregationInterval: aggregationInterval,
 	}
 	collector.clusterAdmin.Set(0)
-	collector.limitedSupport.Set(0)
+	//collector.limitedSupport.Set(0)
```
Comment (fahlmant) [inline pkg/metrics/aggregator.go:94]: Please remember to add this back :)

Comment (aliceh) [reply pkg/metrics/aggregator.go]: :-D

File: pkg/metrics/aggregator.go [hunk 6]
```diff
@@ -90,7 +91,6 @@ func NewMetricsAggregator(aggregationInterval time.Duration) *AdoptionMetricsAgg
 		aggregationInterval: aggregationInterval,
 	}
 	collector.clusterAdmin.Set(0)
-	collector.limitedSupport.Set(0)
```
Comment (fahlmant) [inline pkg/metrics/aggregator.go:93]: Why was this removed?

Comment (aliceh) [reply pkg/metrics/aggregator.go:93]: there is no Set method for GaugeVec if I am not mistaken

Comment (aliceh) [reply pkg/metrics/aggregator.go:93]: I can not answer that. I will run a test as soon as I sort out the errors.

Comment (aliceh) [reply pkg/metrics/aggregator.go:93]: It does seem that we need an initial value, I am looking up in the docs how this can be done.

Comment (aliceh) [reply pkg/metrics/aggregator.go:93]: Actually, upon looking closer, I think our code will set it after checking. We might be Ok.

**Review by mjlshen — CHANGES_REQUESTED** [review]

File: controllers/limited_support/limited_support_controller.go [hunk 7]
```diff
@@ -34,6 +35,8 @@ const (
 	limitedSupportConfigMapNamespace = "openshift-osd-metrics"
 )
 
+const EnvClusterID = "CLUSTER_ID"
```
Comment (mjlshen) [inline controllers/limited_support/limited_support_controller.go:38]: Since this matches https://github.com/openshift/osd-metrics-exporter/blob/d834f20ed0a2314a0bda8b8591dc2b20df1f6251/controllers/proxy/proxy_controller.go#L33

Should we just define the consts in one place? Maybe `controllers/utils/const.go`? Then the proxy controller and this limited support controller could just use `utils.EnvClusterID` and reference the same constant.

File: controllers/limited_support/limited_support_controller.go [hunk 8]
```diff
@@ -49,6 +52,7 @@ func (r *LimitedSupportConfigMapReconciler) Reconcile(ctx context.Context, req c
 	reqLogger.Info("Reconciling Limited Support ConfigMap")
 
 	// Fetch the ConfigMap openshift-osd-metrics/limited-support
+	uuid := os.Getenv(EnvClusterID)
```
Comment (mjlshen) [inline controllers/limited_support/limited_support_controller.go:55]:
```suggestion
	uuid, ok := os.LookupEnv(EnvClusterID)
	if !ok || uuid == "" {
	  return reconcile.Result{}, fmt.Errorf("cluster ID unset or returned as empty string")
	}
```

**Review by mjlshen — CHANGES_REQUESTED** [review]
A couple nits looks good overall!

File: controllers/limited_support/limited_support_controller.go (hunk 7 above)
Comment (mjlshen) [reply controllers/limited_support/limited_support_controller.go:38]: Makes sense, thanks

**Comment by mjlshen** [comment]
I think the last remaining thing is the unit tests, which are failing because they expect the `CLUSTER_ID` environment variable to be set. This can be done by using `t.SetEnv` https://pkg.go.dev/testing#T.Setenv in the corresponding failing tests: https://github.com/openshift/osd-metrics-exporter/blob/d834f20ed0a2314a0bda8b8591dc2b20df1f6251/controllers/limited_support/limited_support_controller_test.go#L27

File: controllers/limited_support/limited_support_controller_test.go [hunk 9]
```diff
@@ -26,19 +27,22 @@ func makeTestConfigMap(name string, namespace string) *corev1.ConfigMap {
...
 			expectedResults: `
 # HELP limited_support_enabled Indicates if limited support is enabled
 # TYPE limited_support_enabled gauge
-limited_support_enabled{name="osd_exporter"} 1
+limited_support_enabled{_id="i-am-a-cluster-id",name="osd_exporter"} 1
 `,
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
+			os.Setenv(EnvClusterID, tc.testEnvClusterID)
```
Comment (mjlshen) [inline controllers/limited_support/limited_support_controller_test.go:45]: Let's use `t.Setenv` here, the difference is small, but `os.Setenv` doesn't clean up after itself, so running these tests locally could start leaking information through environment variables to other tests.

```suggestion
			t.Setenv(EnvClusterID, tc.testEnvClusterID)
```

File: controllers/limited_support/limited_support_controller_test.go [hunk 10]
```diff
@@ -72,19 +76,22 @@ limited_support_enabled{name="osd_exporter"} 1
...
 			expectedResults: `
 # HELP limited_support_enabled Indicates if limited support is enabled
 # TYPE limited_support_enabled gauge
-limited_support_enabled{name="osd_exporter"} 0
+limited_support_enabled{_id="i-am-a-cluster-id",name="osd_exporter"} 0
 `,
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
+			os.Setenv(EnvClusterID, tc.testEnvClusterID)
```
Comment (mjlshen) [inline controllers/limited_support/limited_support_controller_test.go:94]: Same idea here :D Super cool the tests pass now!

```suggestion
			t.Setenv(EnvClusterID, tc.testEnvClusterID)
```

**Review by mjlshen — CHANGES_REQUESTED** [review]

File: controllers/limited_support/limited_support_controller_test.go [hunk 11]
```diff
@@ -2,6 +2,7 @@ package limited_support
 
 import (
 	"context"
+	"os"
```
Comment (sam-nguyen7) [inline controllers/limited_support/limited_support_controller_test.go:5]:
```suggestion
	
```
Can be removed as we're no longer using it

**Review by mjlshen — APPROVED** [review]
/lgtm

**Review by fahlmant — APPROVED** [review]
/approve

**Review by fahlmant — APPROVED** [review]
/approve

**Comment by iamkirkbater** [comment]
/approve

Nice work!
